README.md
# SLM-vcv-rack
plugins for [VCV Rack virtual modular synthesizer](https://vcvrack.com/)

## Mux Freak

![Alt text](screenshots/mux-freak_panel.png?raw=true "Mux Freak")

### short description
clocked 4 to 1 mux with random switching, probability and length controls.

### long description
Each of the 4 channels has a **PROBABILITY** and **LENGTH** control.
The **PROBABILITY** (0% to 100%) determines how likely that channel is going to be selected whenever a random switch will occour
The **LENGTH** (2 to 64 clock cycles) determines how long that channel will play once it had been selected.

After the selected channel has been played for the **LENGTH** number of cycles, a new random mux switch will take place.
On each mux switching a pulse is generated on the **TRIG** output. This can be used to trigger other modules.

To momentarily bypass the mux random switching, a **SOLO** button is available for each channel. This can be handy to audition the channel while racking or in some live situations.

note that the switching will always happens on multiple of the input clock. So it is possibile to sync with master clock wherever desired.

**CVs** to modulate **PROBABILITY** and **LENGTH** are available.

## The Tailor
![Alt text](screenshots/the-tailor_panel.png?raw=true "The Tailor")

### short description
waveforms morphing tool, same concept as the [WTF Module](http://dove-audio.com/wtf-module/) from [Dove Audio](http://dove-audio.com/)

### long description
the tailor will cut a piece of waveform 2 (right side **IN** input) that will replace the corresponding piece from the waveform at left side **IN** input.

the center and the width of the piece to be cut is determined by the **CENTER** and **WIDTH** knobs (and correspoding CVs + attenuerters)

in order for the tailoring to be "musical", the module must operate in sync with one of the 2 waveforms period.

there are 2 methods of synchronization:
1) auto sync (**SYNC** input not connected) - the left side **in** waveform zero crossing rising edge determines the start of the cycle
2) external sync (**SYNC** input connected to a pulse or square waveform) - the start of the cycle is determined by the rising edge of the provided sync signal

In general, for simple waveforms (that is without multiple zero crossings within the cycle) auto sync will give best results, otherwise an external sync signal must be provided.

If ideally your VCO has a corresponding square wave out, you can use that as sync signal. 
However, depending on the actual VCO implementation, the square wave and your actual signal might not be perfectly in phase, especially at higher pitch. 

### additional considerations
#### buffer size
the module has an internal buffer of 100ksamples.

this means that for example, at 48kHz sampling rate, you would buffer overflow for a cycle slower than, roughly, 0.5Hz

when this happens, the red LED below the signal scope will light up to indicate this condition and the module won't produce any output.

This limits somehow the applicability of this module for slow LFOs morphing. However this is mainly intended to operate on VCOs although you can use it however you like, keeping in mind the said limitation.

#### aliasing
the "tailoring" is obviously a highly non-linear process and, depending on the actual waveforms in play, can easily introduce sudden discontinuities in your final signal which in digital domain translates into aliasing.

How much aliasing you can tollerate and methods to mitigate this problem are huge topics on their own. 

Future updates of this module might add optional miniBLEP around the window edges (for example) but at this stage it is behaving like the previusly mentioned eurorack module (that means, no special effort in minimizing aliasing)

### youtube demo
coming soon...

### donate
if you like my mmodules please consider to [donate](https://paypal.me/salvolm)


CHANGELOG.md
### 1.1.0 (2019-07-31)
- New Module: The Tailor (waveforms morphing tool)
- Mux Freak: New Panel, minor code cleanup

### 1.0.0 (2019-07-05)
- First release

src/test.cpp
#include "common.hpp"

struct WaveformPiece {
    float buffer[10 * 1024] = {0.0f};
    int size;
    float strokeWidth;
    NVGcolor strokeColor;
};

struct Test : Module {
	enum ParamId {
        WINDOW_CENTER_PARAM,
        WINDOW_WIDTH_PARAM,
        WINDOW_CENTER_ATTENUERTER,
        WINDOW_WIDTH_ATTENUERTER,
		NUM_PARAMS
	};
	enum InputId {
        CH1_INPUT,
        CH2_INPUT,
        SYNC_INPUT,
        WINDOW_CENTER_CV,
        WINDOW_WIDTH_CV,
		NUM_INPUTS
	};
	enum OutputId {
        MAIN_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightId {
		NUM_LIGHTS
	};

    int sampleCounter = 0;
    int samplesPerPeriod = 0;
    int syncCount = 0;

    //struct WaveformPiece waveform[3];

    dsp::SchmittTrigger schmittTrigger;
    bool readyToDisplay = false;
    
    float displayBuffer[20 * 1024] = {0.0f};

	Test() {
		// Configure the module
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        configParam(WINDOW_CENTER_PARAM, 0.0f, 1.0f, 0.5f, "Window Center");
        configParam(WINDOW_WIDTH_PARAM, 0.0f, 1.0f, 0.3f, "Window Width");
	}

	void process(const ProcessArgs &args) override {

        if (!inputs[SYNC_INPUT].isConnected())
            return;

        if (schmittTrigger.process(rescale(inputs[SYNC_INPUT].getVoltage(), 0.1f, 2.f, 0.f, 1.f))) {
            samplesPerPeriod = sampleCounter;
            sampleCounter = 0;
            
            readyToDisplay = true;
        }

        float samplePosition = (float)sampleCounter / (float)samplesPerPeriod;
        
        float windowCenterCv = inputs[WINDOW_CENTER_CV].getVoltage() * params[WINDOW_CENTER_ATTENUERTER].getValue();
        float windowWidthCv = inputs[WINDOW_WIDTH_CV].getVoltage() * params[WINDOW_WIDTH_ATTENUERTER].getValue();
        
        float windowCenter = params[WINDOW_CENTER_PARAM].getValue() + windowCenterCv;
        float windowWidth = params[WINDOW_WIDTH_PARAM].getValue() + windowWidthCv;

        windowCenter = clamp(windowCenter, 0.0f, 1.0f);
        windowWidth = clamp(windowWidth, 0.0f, 1.0f);
 
        if (samplePosition >= (windowCenter - windowWidth / 2.0f) && samplePosition <= (windowCenter + windowWidth / 2.0f)) {
            outputs[MAIN_OUTPUT].setVoltage(inputs[CH2_INPUT].getVoltage());
        } else {
            outputs[MAIN_OUTPUT].setVoltage(inputs[CH1_INPUT].getVoltage());
        }

        displayBuffer[sampleCounter++] = outputs[MAIN_OUTPUT].getVoltage();

	}
};

struct TestDisplay : TransparentWidget {
	Test *module;

    void draw(const DrawArgs &args) override {
        if (!module)
			return;

        if (!module->readyToDisplay)
            return;

        int size = module->samplesPerPeriod;
        
        float width = box.size.x;
        float deltaTime = width / (float)size;


        float max = -10000000.0;
        float min = +10000000.0;

        for (int i = 0; i < size; i++)
        {
            if (module->displayBuffer[i] > max)
            {
                max = module->displayBuffer[i];
            }

            if (module->displayBuffer[i] < min)
            {
                min = module->displayBuffer[i];
            }
        }

        // float drawHeight = 100.0f;
        float center = box.size.y / 2.0f;
        float gain = - 8.0f;
        
        nvgSave(args.vg);
        nvgBeginPath(args.vg);

        nvgMoveTo(args.vg, 0, center + gain * module->displayBuffer[0]);
        
        for (int i = 0; i < size; i++) {
            nvgLineTo(args.vg, (float)i * deltaTime, center + gain * module->displayBuffer[i]);
        }

        nvgStrokeWidth(args.vg, 1.0f);
        nvgStrokeColor(args.vg, nvgRGBA(0xFF, 0xFF, 0xFF, 0xFF));
        nvgStroke(args.vg);

        nvgRestore(args.vg);

        
        module->readyToDisplay = false;
    }
};



struct TestWidget : ModuleWidget {
	TestWidget(Test *module) {
		setModule(module);
		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/test.svg")));

        {
			TestDisplay *display = new TestDisplay();
			display->module = module;
			display->box.pos = Vec(0, 44);
			display->box.size = Vec(box.size.x, 140);
			addChild(display);
		}

        addParam(createParam<SLMSmallKnobGreen>(Vec(15, 270), module, Test::WINDOW_CENTER_PARAM));
        addParam(createParam<SLMSmallKnobGreen>(Vec(15, 300), module, Test::WINDOW_CENTER_ATTENUERTER));
        addInput(createInput<SLMInputPort>(Vec(15, 350), module, Test::WINDOW_CENTER_CV));
        addParam(createParam<SLMSmallKnobGreen>(Vec(60, 270), module, Test::WINDOW_WIDTH_PARAM));
        addParam(createParam<SLMSmallKnobGreen>(Vec(60, 300), module, Test::WINDOW_WIDTH_ATTENUERTER));
        addInput(createInput<SLMInputPort>(Vec(60, 350), module, Test::WINDOW_WIDTH_CV));
        addInput(createInput<SLMInputPort>(Vec(17, 185), module, Test::CH1_INPUT));
        addInput(createInput<SLMInputPort>(Vec(57, 185), module, Test::CH2_INPUT));
        addInput(createInput<SLMInputPort>(Vec(17, 225), module, Test::SYNC_INPUT));

        addOutput(createOutput<SLMOutputPort>(Vec(129, 327), module, Test::MAIN_OUTPUT));
	}
};


// Define the Model with the Module type, ModuleWidget type, and module slug
Model *modelTest = createModel<Test, TestWidget>("test");


src/common.hpp
#ifndef COMMON_H
#define COMMON_H

#include "plugin.hpp"

struct SLMTrimmer : app::SvgKnob {
	SLMTrimmer() {
		minAngle = -0.67*M_PI;
		maxAngle = 0.67*M_PI;
		//snap = true;
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/trimmer.svg")));
	}
};

struct SLMSmallKnobGreen : app::SvgKnob {
	SLMSmallKnobGreen() {
		minAngle = -0.67*M_PI;
		maxAngle = 0.67*M_PI;
		//snap = true;
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/knob-green.svg")));
	}
};

struct SLMBigKnobGreen : app::SvgKnob {
	SLMBigKnobGreen() {
		minAngle = -0.67*M_PI;
		maxAngle = 0.67*M_PI;
		//snap = true;
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/knob-green-big.svg")));
	}
};

struct SLMLight : GrayModuleLightWidget {
	SLMLight(unsigned char r, unsigned char g, unsigned char b) {
		addBaseColor(nvgRGB(r, g, b));
	}
};

struct SLMRedLight : SLMLight {
	SLMRedLight() : SLMLight(0xFE, 0x71, 0x7D) {}
};

struct SLMSmallKnobRed : app::SvgKnob {
	SLMSmallKnobRed() {
		minAngle = -0.67*M_PI;
		maxAngle =  0.67*M_PI;
		//snap = true;
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/knob-red.svg")));
	}
};

struct SLMBigKnobRed : app::SvgKnob {
	SLMBigKnobRed() {
		minAngle = -0.67*M_PI;
		maxAngle =  0.67*M_PI;
		//snap = true;
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/knob-red-big.svg")));
	}
};

struct SLMInputPort : app::SvgPort {
	SLMInputPort() {
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/port.svg")));
	}
};

struct SLMOutputPort : app::SvgPort {
	SLMOutputPort() {
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/port.svg")));
	}
};

struct SLMToggleButton : app::SvgSwitch {
	SLMToggleButton() {
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/button-off.svg")));
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/button-on.svg")));
	}
};

struct SLMScrew : app::SvgScrew {
	SLMScrew() {
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/screw.svg")));
	}
};

#endif // COMMON_H


src/mux-freak.cpp
#include "common.hpp"
#include <ctime>

struct MuxFreak : Module {
	enum ParamId {
		ENUMS(LEN_PARAM, 4),
		ENUMS(PROB_PARAM, 4),
		ENUMS(SOLO_PARAM, 4),
		NUM_PARAMS
	};
	enum InputId {
		CLK_INPUT,
		ENUMS(CH_INPUT, 4),
		ENUMS(LEN_INPUT, 4),
		ENUMS(PROB_INPUT, 4),
		NUM_INPUTS
	};
	enum OutputId {
		TRIG_OUTPUT,
		MUX_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightId {
		NUM_LIGHTS
	};

	float relativeProb[4];
	int playingNow = 0;
	int remainingTicks = 1;
	float clockBefore = 0.0f;
	int soloIndex = -1;

	dsp::PulseGenerator pulseGenerator;
	dsp::BooleanTrigger buttonTrigger[4];

	MuxFreak() {
		// Configure the module
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);

		// Configure parameters
		// See engine/Param.hpp for config() arguments
		// void configParam(int paramId, float minValue, float maxValue, float defaultValue, std::string label = "", std::string unit = "", float displayBase = 0.f, float displayMultiplier = 1.f, float displayOffset = 0.f)
		configParam(SOLO_PARAM + 0, 0.0f, 1.0f, 0.0f, "CH1 Solo");
		configParam(SOLO_PARAM + 1, 0.0f, 1.0f, 0.0f, "CH2 Solo");
		configParam(SOLO_PARAM + 2, 0.0f, 1.0f, 0.0f, "CH3 Solo");
		configParam(SOLO_PARAM + 3, 0.0f, 1.0f, 0.0f, "CH4 Solo");
		
		configParam(LEN_PARAM + 0, 2.0f, 64.0f, 8.0f, "CH1 Length", " clock counts");
		configParam(LEN_PARAM + 1, 2.0f, 64.0f, 8.0f, "CH2 Length", " clock counts");
		configParam(LEN_PARAM + 2, 2.0f, 64.0f, 8.0f, "CH3 Length", " clock counts");
		configParam(LEN_PARAM + 3, 2.0f, 64.0f, 8.0f, "CH4 Length", " clock counts");

		configParam(PROB_PARAM + 0, 0.0f, 100.0f, 25.0f, "CH1 Probability", "%");
		configParam(PROB_PARAM + 1, 0.0f, 100.0f, 25.0f, "CH2 Probability", "%");
		configParam(PROB_PARAM + 2, 0.0f, 100.0f, 25.0f, "CH3 Probability", "%");
		configParam(PROB_PARAM + 3, 0.0f, 100.0f, 25.0f, "CH4 Probability", "%");

		srand (static_cast <unsigned> (time(0)));
	}

	int coinToss()
	{
		if (soloIndex >= 0)
		{
			return soloIndex;
		}

		// calculate the relative probabilities by filtering out not active channels
		float probSum = 0;
		for (int i = 0; i < 4; i++)
		{
			if (inputs[CH_INPUT + i].isConnected())
				probSum += 0.01f * clamp(params[PROB_PARAM + i].getValue() + inputs[PROB_INPUT + i].getVoltage(), 0.0f, 100.0f);
		}

		for (int i = 0; i < 4; i++)
		{
			if (inputs[CH_INPUT + i].isConnected())
			{
				relativeProb[i] = 0.01f * clamp(params[PROB_PARAM + i].getValue() + inputs[PROB_INPUT + i].getVoltage(), 0.0f, 100.0f) / probSum;
			}
			else
			{
				relativeProb[i] = 0.0f;
			}
				
		}

		// coin toss
		float coin = (float)rand() / RAND_MAX;

		float p0 = relativeProb[0];
		float p1 = p0 + relativeProb[1];
		float p2 = p1 + relativeProb[2];

		if (coin <= p0)
		{
			return 0;
		}

		if (coin > p0 && coin <= p1)
		{
			return 1;
		}

		if (coin > p1 && coin <= p2)
		{
			return 2;
		}

		return 3;
	}

	void process(const ProcessArgs &args) override {

		outputs[TRIG_OUTPUT].setVoltage(pulseGenerator.process(args.sampleTime) ? 10.0f : 0.0f);

		// check solo buttons
		for (int i = 0; i < 4; i++)
		{
			bool triggered = buttonTrigger[i].process(params[SOLO_PARAM + i].getValue() > 0.0f);
			if (triggered)
			{
				for (int j = 0; j < 4; j++)
				{
					if (j != i)
					{
						params[SOLO_PARAM + j].setValue(0.0f);
					}
				}
			}
		}

		for (int i = 0; i < 4; i++)
		{
			if (params[SOLO_PARAM + i].getValue() > 0.0f)
			{
				soloIndex = i;
				break;
			}
			soloIndex = -1;
		}

		// no clock or no active inputs -> nothing to do!
		if (!inputs[CLK_INPUT].isConnected() || (!inputs[CH_INPUT + 0].isConnected() && 
		!inputs[CH_INPUT + 1].isConnected() && 
		!inputs[CH_INPUT + 2].isConnected() &&
		!inputs[CH_INPUT + 3].isConnected()))
		{
			return;
		}

		// detect clock rising edge
		float clockNow = inputs[CLK_INPUT].getVoltage();
		
		if (clockNow - clockBefore > 0.5f)
		{
			remainingTicks--;
			if (remainingTicks == 0) // time for a new coin toss!
			{
				playingNow = coinToss();
				remainingTicks = (int)round(clamp(params[LEN_PARAM + playingNow].getValue() + inputs[LEN_INPUT + playingNow].getVoltage(), 2.0f, 64.0f));
				pulseGenerator.trigger();
			}
		}
		
		clockBefore = clockNow;
		outputs[MUX_OUTPUT].setVoltage(inputs[CH_INPUT + playingNow].getVoltage());
		
	}
};


struct MuxFreakWidget : ModuleWidget {
	MuxFreakWidget(MuxFreak *module) {
		setModule(module);
		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/mux-freak.svg")));

		addChild(createWidget<SLMScrew>(Vec(RACK_GRID_WIDTH, 0.f)));
		addChild(createWidget<SLMScrew>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
		addChild(createWidget<SLMScrew>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
		addChild(createWidget<SLMScrew>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

		addParam(createParam<SLMToggleButton>(Vec(20, 56), module, MuxFreak::SOLO_PARAM + 0));
		addParam(createParam<SLMToggleButton>(Vec(65, 56), module, MuxFreak::SOLO_PARAM + 1));
		addParam(createParam<SLMToggleButton>(Vec(110, 56), module, MuxFreak::SOLO_PARAM + 2));
		addParam(createParam<SLMToggleButton>(Vec(155, 56), module, MuxFreak::SOLO_PARAM + 3));

		addParam(createParam<SLMSmallKnobGreen>(Vec(15, 103), module, MuxFreak::PROB_PARAM + 0));
		addParam(createParam<SLMSmallKnobGreen>(Vec(60, 103), module, MuxFreak::PROB_PARAM + 1));
		addParam(createParam<SLMSmallKnobGreen>(Vec(105, 103), module, MuxFreak::PROB_PARAM + 2));
		addParam(createParam<SLMSmallKnobGreen>(Vec(150, 103), module, MuxFreak::PROB_PARAM + 3));

		SLMSmallKnobRed* redKnobs[4];
		redKnobs[0] = createParam<SLMSmallKnobRed>(Vec(15, 160), module, MuxFreak::LEN_PARAM + 0);
		redKnobs[1] = createParam<SLMSmallKnobRed>(Vec(60, 160), module, MuxFreak::LEN_PARAM + 1);
		redKnobs[2] = createParam<SLMSmallKnobRed>(Vec(105, 160), module, MuxFreak::LEN_PARAM + 2);
		redKnobs[3] = createParam<SLMSmallKnobRed>(Vec(150, 160), module, MuxFreak::LEN_PARAM + 3);

		for (int i = 0; i < 4; i++) {
			redKnobs[i]->snap = true;
			addParam(redKnobs[i]);
		}

		addInput(createInput<SLMInputPort>(Vec(49, 228), module, MuxFreak::CH_INPUT + 0));
		addInput(createInput<SLMInputPort>(Vec(49, 260), module, MuxFreak::CH_INPUT + 1));
		addInput(createInput<SLMInputPort>(Vec(49, 292), module, MuxFreak::CH_INPUT + 2));
		addInput(createInput<SLMInputPort>(Vec(49, 324), module, MuxFreak::CH_INPUT + 3));

		addInput(createInput<SLMInputPort>(Vec(85, 228), module, MuxFreak::PROB_INPUT + 0));
		addInput(createInput<SLMInputPort>(Vec(85, 260), module, MuxFreak::PROB_INPUT + 1));
		addInput(createInput<SLMInputPort>(Vec(85, 292), module, MuxFreak::PROB_INPUT + 2));
		addInput(createInput<SLMInputPort>(Vec(85, 324), module, MuxFreak::PROB_INPUT + 3));

		addInput(createInput<SLMInputPort>(Vec(120, 228), module, MuxFreak::LEN_INPUT + 0));
		addInput(createInput<SLMInputPort>(Vec(120, 260), module, MuxFreak::LEN_INPUT + 1));
		addInput(createInput<SLMInputPort>(Vec(120, 292), module, MuxFreak::LEN_INPUT + 2));
		addInput(createInput<SLMInputPort>(Vec(120, 324), module, MuxFreak::LEN_INPUT + 3));

		addInput(createInput<SLMInputPort>(Vec(17, 276), module, MuxFreak::CLK_INPUT));

		addOutput(createOutput<SLMOutputPort>(Vec(161, 260), module, MuxFreak::TRIG_OUTPUT));
		addOutput(createOutput<SLMOutputPort>(Vec(161, 291), module, MuxFreak::MUX_OUTPUT));
	}
};


// Define the Model with the Module type, ModuleWidget type, and module slug
Model *modelMuxFreak = createModel<MuxFreak, MuxFreakWidget>("mux-freak");


src/the-tailor.cpp
#include "common.hpp"

#define BUFFER_SIZE 100 * 1024

struct TheTailor : Module {
	enum ParamId {
        WINDOW_CENTER_PARAM,
        WINDOW_WIDTH_PARAM,
        WINDOW_CENTER_ATTENUERTER,
        WINDOW_WIDTH_ATTENUERTER,
		NUM_PARAMS
	};
	enum InputId {
        CH1_INPUT,
        CH2_INPUT,
        SYNC_INPUT,
        WINDOW_CENTER_CV,
        WINDOW_WIDTH_CV,
		NUM_INPUTS
	};
	enum OutputId {
        MAIN_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightId {
        OVERFLOW_ALERT_LIGHT,
		NUM_LIGHTS
	};

    int sampleCounter = 0;
    int samplesPerPeriod = 0;
    int syncCount = 0;

    bool previousAutoSyncLevel = false;

    dsp::SchmittTrigger schmittTrigger;
    bool readyToDisplay = false;
    
    float displayBuffer[BUFFER_SIZE] = {0.0f};

	TheTailor() {
		// Configure the module
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        configParam(WINDOW_CENTER_PARAM, 0.0f, 1.0f, 0.5f, "Window Center");
        configParam(WINDOW_WIDTH_PARAM, 0.0f, 1.0f, 0.3f, "Window Width");
        configParam(WINDOW_CENTER_ATTENUERTER, -1.0f, 1.0f, 0.0f, "Window Center CV Attenuerter");
        configParam(WINDOW_WIDTH_ATTENUERTER, -1.0f, 1.0f, 0.0f, "Window Width CV Attenuerter");
	}

	void process(const ProcessArgs &args) override {

        bool sync = false;

        if (!inputs[SYNC_INPUT].isConnected()) {
            // we are using auto sync
            bool currentAutoSyncLevel = inputs[CH1_INPUT].getVoltage() > 0.0f;
            sync = currentAutoSyncLevel && (!previousAutoSyncLevel);
            previousAutoSyncLevel = currentAutoSyncLevel;
        } else {
            sync = schmittTrigger.process(rescale(inputs[SYNC_INPUT].getVoltage(), 0.1f, 2.f, 0.f, 1.f));
        }

        if (sync) {
            samplesPerPeriod = sampleCounter;
            sampleCounter = 0;
            readyToDisplay = true;
        }

        float samplePosition = (float)sampleCounter / (float)samplesPerPeriod;
        
        float windowCenterCv = inputs[WINDOW_CENTER_CV].getVoltage() * params[WINDOW_CENTER_ATTENUERTER].getValue();
        float windowWidthCv = inputs[WINDOW_WIDTH_CV].getVoltage() * params[WINDOW_WIDTH_ATTENUERTER].getValue();
        
        float windowCenter = params[WINDOW_CENTER_PARAM].getValue() + windowCenterCv;
        float windowWidth = params[WINDOW_WIDTH_PARAM].getValue() + windowWidthCv;

        windowCenter = clamp(windowCenter, 0.0f, 1.0f);
        windowWidth = clamp(windowWidth, 0.0f, 1.0f);
 
        if (samplePosition >= (windowCenter - windowWidth / 2.0f) && samplePosition <= (windowCenter + windowWidth / 2.0f)) {
            outputs[MAIN_OUTPUT].setVoltage(inputs[CH2_INPUT].getVoltage());
        } else {
            outputs[MAIN_OUTPUT].setVoltage(inputs[CH1_INPUT].getVoltage());
        }

        if (sampleCounter < BUFFER_SIZE) {
            displayBuffer[sampleCounter++] = outputs[MAIN_OUTPUT].getVoltage();
        }

        lights[OVERFLOW_ALERT_LIGHT].setBrightness(sampleCounter < BUFFER_SIZE ? 0.0f : 1.0f);

	}
};

struct TheTailorDisplay : TransparentWidget {
	TheTailor *module;

    void draw(const DrawArgs &args) override {
        if (!module)
			return;

        if (!module->readyToDisplay)
            return;

        int size = module->samplesPerPeriod;
        
        float width = box.size.x;
        float deltaTime = width / (float)size;

        // float drawHeight = 100.0f;
        float center = box.size.y / 2.0f;
        float gain = - 4.0f;
        
        nvgSave(args.vg);
        nvgBeginPath(args.vg);

        nvgMoveTo(args.vg, 0, center + gain * module->displayBuffer[0]);
        
        for (int i = 0; i < size; i++) {
            nvgLineTo(args.vg, (float)i * deltaTime, center + gain * module->displayBuffer[i]);
        }

        nvgStrokeWidth(args.vg, 1.0f);
        nvgStrokeColor(args.vg, nvgRGBA(0xEB, 0xEB, 0xEB, 0xFF));
        nvgStroke(args.vg);

        nvgRestore(args.vg);

        module->readyToDisplay = false;
    }
};



struct TheTailorWidget : ModuleWidget {
	TheTailorWidget(TheTailor *module) {
		setModule(module);
		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/the-tailor.svg")));

        addChild(createWidget<SLMScrew>(Vec(RACK_GRID_WIDTH, 0.f)));
		addChild(createWidget<SLMScrew>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
		addChild(createWidget<SLMScrew>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
		addChild(createWidget<SLMScrew>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

        {
			TheTailorDisplay *display = new TheTailorDisplay();
			display->module = module;
			display->box.pos = Vec(10, 38);
			display->box.size = Vec(130, 76);
			addChild(display);
		}

        addParam(createParam<SLMBigKnobGreen>(Vec(19, 138), module, TheTailor::WINDOW_CENTER_PARAM));
        addParam(createParam<SLMTrimmer>(Vec(10, 190), module, TheTailor::WINDOW_CENTER_ATTENUERTER));
        addInput(createInput<SLMInputPort>(Vec(19, 272), module, TheTailor::WINDOW_CENTER_CV));

        addParam(createParam<SLMBigKnobRed>(Vec(93, 138), module, TheTailor::WINDOW_WIDTH_PARAM));
        addParam(createParam<SLMTrimmer>(Vec(116, 190), module, TheTailor::WINDOW_WIDTH_ATTENUERTER));
        addInput(createInput<SLMInputPort>(Vec(105, 272), module, TheTailor::WINDOW_WIDTH_CV));

        addInput(createInput<SLMInputPort>(Vec(39, 228), module, TheTailor::CH1_INPUT));
        addInput(createInput<SLMInputPort>(Vec(85, 228), module, TheTailor::CH2_INPUT));

        addInput(createInput<SLMInputPort>(Vec(62, 272), module, TheTailor::SYNC_INPUT));

        addOutput(createOutput<SLMOutputPort>(Vec(62, 310), module, TheTailor::MAIN_OUTPUT));
        addChild(createLight<MediumLight<SLMRedLight>>(Vec(71, 133), module, TheTailor::OVERFLOW_ALERT_LIGHT));
	}
};


// Define the Model with the Module type, ModuleWidget type, and module slug
Model *modelTheTailor = createModel<TheTailor, TheTailorWidget>("the-tailor");
